File: matiasmasso/Mat.NET/DT2/DTOContestBase.vb
Public Class DTOContestBase
    Inherits DTOBaseGuid
    Property Lang As DTOLang
    Property Country As DTOCountry
    Property Title As String = ""
    Property Bases As String
    Property FchFrom As Date = Date.MinValue
    Property FchTo As Date = Date.MinValue
    Property UrlExterna As String
    Property CostPrize As DTOAmt
    Property CostPubli As DTOAmt

    <JsonIgnore> Property ImageFbFeatured As Image
    <JsonIgnore> Property ImageBanner600 As Image
    <JsonIgnore> Property ImageCallToAction500 As Image
    Property Visible As Boolean

    Property Codi As Codis

    Public Enum Codis
        NotSet
        Raffle
        Contest
    End Enum

    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

    Public Function ImageFbFeaturedUrl(Optional AbsoluteUrl As Boolean = False) As String
        Dim retval = MmoUrl.image(Defaults.ImgTypes.sorteofbfeatured200, MyBase.Guid, AbsoluteUrl)
        Return retval
    End Function

    Shared Function IsNotYetStarted(oContestBase As DTOContestBase) As Boolean
        Dim retval As Boolean = True
        If IsOver(oContestBase) Then
            retval = False
        ElseIf IsActive(oContestBase) Then
            retval = False
        End If
        Return retval
    End Function

    Public Function IsOver() As Boolean
        Dim retval As Boolean = _FchTo < Now
        Return retval
    End Function

    Shared Function IsOver(oContestBase As DTOContestBase) As Boolean
        Dim retval As Boolean
        If oContestBase IsNot Nothing Then
            retval = oContestBase.IsOver
        End If
        Return retval
    End Function

    Shared Function IsActive(oContestBase As DTOContestBase) As Boolean
        Dim retval As Boolean
        If IsOver(oContestBase) Then
            retval = False
        Else
            Select Case oContestBase.FchFrom.Year
                Case Is > Now.Year
                    retval = False
                Case Is < Now.Year
                    retval = True
                Case Is = Now.Year
                    Select Case oContestBase.FchFrom.Month
                        Case Is > Now.Month
                            retval = False
                        Case Is < Now.Month
                            retval = True
                        Case Is = Now.Month
                            Select Case oContestBase.FchFrom.Day
                                Case Is > Now.Day
                                    retval = False
                                Case Is < Now.Day
                                    retval = True
                                Case Is = Now.Day
                                    retval = True
                            End Select
                    End Select
            End Select
        End If

        Return retval
    End Function



End Class


Public Class DTOContestBaseParticipant
    Inherits DTOBaseGuid

    Property Parent As DTOContestBase
    Property Fch As Date
    Property User As DTOUser

    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

End Class

File: matiasmasso/Mat.NET/DTO/DTOContestBase.cs
using DTO;
using Newtonsoft.Json;
using SixLabors.ImageSharp;
using System;

public class DTOContestBase : DTOBaseGuid
{
    public DTOLang Lang { get; set; }
    public DTOCountry Country { get; set; }
    public string Title { get; set; } = "";
    public string Bases { get; set; }
    public DateTime FchFrom { get; set; } = DateTime.MinValue;
    public DateTime FchTo { get; set; } = DateTime.MinValue;
    public string UrlExterna { get; set; }
    public DTOAmt CostPrize { get; set; }
    public DTOAmt CostPubli { get; set; }

    [JsonIgnore]
    public Image ImageFbFeatured { get; set; }
    [JsonIgnore]
    public Image ImageBanner600 { get; set; }
    [JsonIgnore]
    public Image ImageCallToAction500 { get; set; }
    public bool Visible { get; set; }

    public Codis Codi { get; set; }

    public enum Codis
    {
        NotSet,
        Raffle,
        Contest
    }

    public DTOContestBase() : base()
    {
    }

    public DTOContestBase(Guid oGuid) : base(oGuid)
    {
    }

    public string ImageFbFeaturedUrl(bool AbsoluteUrl = false)
    {
        var retval = MmoUrl.image(Defaults.ImgTypes.sorteofbfeatured200, base.Guid, AbsoluteUrl);
        return retval;
    }

    public static bool IsNotYetStarted(DTOContestBase oContestBase)
    {
        bool retval = true;
        if (IsOver(oContestBase))
            retval = false;
        else if (IsActive(oContestBase))
            retval = false;
        return retval;
    }

    public bool IsOver()
    {
        bool retval = FchTo < DateTime.Now;
        return retval;
    }

    public static bool IsOver(DTOContestBase oContestBase)
    {
        bool retval = false;
        if (oContestBase != null)
            retval = oContestBase.IsOver();
        return retval;
    }

    public static bool IsActive(DTOContestBase oContestBase)
    {
        bool retval = false;
        if (IsOver(oContestBase))
            retval = false;
        else
            switch (oContestBase.FchFrom.Year)
            {
                case object _ when oContestBase.FchFrom.Year > DateTime.Now.Year:
                    {
                        retval = false;
                        break;
                    }

                case object _ when oContestBase.FchFrom.Year < DateTime.Now.Year:
                    {
                        retval = true;
                        break;
                    }

                case object _ when oContestBase.FchFrom.Year == DateTime.Now.Year:
                    {
                        switch (oContestBase.FchFrom.Month)
                        {
                            case object _ when oContestBase.FchFrom.Month > DateTime.Now.Month:
                                {
                                    retval = false;
                                    break;
                                }

                            case object _ when oContestBase.FchFrom.Month < DateTime.Now.Month:
                                {
                                    retval = true;
                                    break;
                                }

                            case object _ when oContestBase.FchFrom.Month == DateTime.Now.Month:
                                {
                                    switch (oContestBase.FchFrom.Day)
                                    {
                                        case object _ when oContestBase.FchFrom.Day > DateTime.Now.Day:
                                            {
                                                retval = false;
                                                break;
                                            }

                                        case object _ when oContestBase.FchFrom.Day < DateTime.Now.Day:
                                            {
                                                retval = true;
                                                break;
                                            }

                                        case object _ when oContestBase.FchFrom.Day == DateTime.Now.Day:
                                            {
                                                retval = true;
                                                break;
                                            }
                                    }

                                    break;
                                }
                        }

                        break;
                    }
            }

        return retval;
    }
}


public class DTOContestBaseParticipant : DTOBaseGuid
{


    public DTOContestBase Parent { get; set; }
    public DateTime Fch { get; set; }
    public DTOUser User { get; set; }

    public DTOContestBaseParticipant() : base()
    {
    }

    public DTOContestBaseParticipant(Guid oGuid) : base(oGuid)
    {
    }
}